### PR TITLE
SAMloader

### PR DESCRIPTION
**Pull request by KitVB**
## added SAMLoader

Added featurizer, tests, sample files and the loader

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [x] Run `mypy -p deepchem` and check no errors
  - [x] Run `flake8 <modified file> --count` and check no errors
  - [x] Run `python -m doctest <modified file>` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/data/data_loader.py [hunk 1]
```diff
@@ -11,12 +11,14 @@
 
 import pandas as pd
 import numpy as np
+import pysam
```
Comment (shreyasvinaya) [inline deepchem/data/data_loader.py:14]: add this to a try except block

File: deepchem/feat/bio_seq_featurizer.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
+import numpy as np
+import deepchem as dc
+import pysam
+
+
+class SAMFeaturizer:
```
Comment (shreyasvinaya) [inline deepchem/feat/bio_seq_featurizer.py:6]: add docstrings

File: deepchem/feat/bio_seq_featurizer.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+    def __init__(self, max_records=None):
+        self.max_records = max_records
+
+    def get_features(self, samfile):
+        """
```
Comment (shreyasvinaya) [inline deepchem/feat/bio_seq_featurizer.py:12]: use [Numpy style](https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard) docstrings

File: deepchem/feat/bio_seq_featurizer.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
+import numpy as np
+import deepchem as dc
+import pysam
```
Comment (shreyasvinaya) [inline deepchem/feat/bio_seq_featurizer.py:3]: add pysam to the testing environment `env_common.yml`

File: deepchem/data/data_loader.py [hunk 5]
```diff
@@ -1753,3 +1755,67 @@ def _featurize_shard(self, shard):
...
+    Sequence Alignment Map (SAM) is a text-based format used to store 
+    biological sequences aligned to a reference sequence. This
+    class provides convenience files to lead SAM data and
+    extract their features for use in downstream
+    learning tasks.
+    """
+
+    def __init__(self, featurizer: Optional[Featurizer] = None):
+        """Initialize SAMLoader.
+        Parameters
```
Comment (shreyasvinaya) [inline deepchem/data/data_loader.py:1771]: add new line before the start of parameters, [Numpy style](https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard) docstrings

File: deepchem/data/data_loader.py [hunk 6]
```diff
@@ -1753,3 +1755,67 @@ def _featurize_shard(self, shard):
...
+
+        # Set self.featurizer
+        self.featurizer = featurizer
+
+    def create_dataset(self,
+                       input_files: OneOrMany[str],
+                       data_dir: Optional[str] = None,
+                       shard_size: Optional[int] = None) -> DiskDataset:
+        """Creates a `Dataset` from input SAM files.
+        Parameters
```
Comment (shreyasvinaya) [inline deepchem/data/data_loader.py:1795]: add new line before the start of parameters

**Review by shreyasvinaya — CHANGES_REQUESTED** [review]
some fixes required

File: deepchem/data/data_loader.py [hunk 7]
```diff
@@ -1753,3 +1755,67 @@ def _featurize_shard(self, shard):
...
+
+
+class SAMLoader(DataLoader):
+    """Handles loading of SAM files.
+    Sequence Alignment Map (SAM) is a text-based format used to store 
+    biological sequences aligned to a reference sequence. This
+    class provides convenience files to lead SAM data and
+    extract their features for use in downstream
+    learning tasks.
+    """
```
Comment (shreyasvinaya) [inline deepchem/data/data_loader.py:1767]: add usage examples

File: deepchem/feat/bio_seq_featurizer.py [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+        The maximum number of records to extract from the SAM file. If None, all records will be extracted.
+
+    Attributes
+    ----------
+    max_records : int or None
+        The maximum number of records to extract.
+
+    Methods
+    -------
+    get_features(samfile)
```
Comment (rbharath) [inline deepchem/feat/bio_seq_featurizer.py:54]: Should be removed

File: deepchem/feat/bio_seq_featurizer.py [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+try:
+    import pysam
+except ImportError:
+    print("Error: Unable to import pysam. Please make sure it is installed.")
+from deepchem.feat import Featurizer
+
+class SAMFeaturizer(Featurizer):
+    """
+    This class extracts Query Name, Query Sequence, Query Length, Reference Name, 
+    Reference Start, CIGAR and Mapping Quality of the alignment in the SAM file.
```
Comment (rbharath) [inline deepchem/feat/bio_seq_featurizer.py:12]: Expand the docstring here. What is the type of input, what is the type of output?

File: deepchem/data/tests/test_sam_loader.py [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+        """
+        sam_featurizer = dc.feat.SAMFeaturizer(max_records=5)
+        sam_file_path = os.path.join(self.current_dir, "example.sam")
+        samfile = pysam.AlignmentFile(sam_file_path, "r")
+        dataset = sam_featurizer._featurize(samfile)
+
+        assert dataset.shape == (5, 7)
+
+
+if __name__ == "__main__":
```
Comment (rbharath) [inline deepchem/data/tests/test_sam_loader.py:55]: Let's remove this not needed

File: deepchem/data/data_loader.py [hunk 11]
```diff
@@ -1858,3 +1865,83 @@ def _featurize_shard(self, shard):
...
+
+
+class SAMLoader(DataLoader):
+    """Handles loading of SAM files.
+
+    Sequence Alignment Map (SAM) is a text-based format used for storing biological sequences 
+    aligned to a reference sequence.It is generally used for storing nucleotide sequences, 
+    generated by next generation sequencing technologies, and unmapped sequences.
+
+    This class provides methods to load and featurize data from SAM files.
```
Comment (rbharath) [inline deepchem/data/data_loader.py:1877]: Add a lot more docs. Explain how the SAM format works and provide examples

File: deepchem/feat/bio_seq_featurizer.py [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+
+    Examples
+    --------
+    >>> from deepchem.data.data_loader import SAMLoader
+    >>> import deepchem as dc
+    >>> inputs = 'deepchem/data/tests/example.sam'
+    >>> featurizer = dc.feat.SAMFeaturizer()
+    >>> features = featurizer.featurize(inputs)
```
Comment (rbharath) [inline deepchem/feat/bio_seq_featurizer.py:20]: A featurizer should be RAM to RAM and not touch disk. Let's have the featurizer accept a string buffer or some other in memory object

**Review by rbharath — COMMENTED** [review]
Let's split this into two files, one for the featurizer and one for the loader

**Pull request closed without merging**